crates/sema/src/ty/ty.rs

```diff
@@ -451,6 +451,16 @@ impl<'gcx> Ty<'gcx> {
                 }
             }
 
+            // Contract -> Base contract (inheritance check)
+            (Contract(self_contract_id), Contract(other_contract_id)) => {
+                let self_contract = gcx.hir.contract(self_contract_id);
+                if self_contract.linearized_bases.contains(&other_contract_id) {
+                    Ok(())
+                } else {
+                    Result::Err(())
```

Comment:
I think we can have an `enum TyConvertError { Generic, ... }` that gives more context about the conversion failure, like in this case it can be `NonDerivedContract => "{a} does not inherit from {b}"`

solc doesn't give this info tho
```solidity
Error: Type contract Unrelated is not implicitly convertible to expected type contract Base.
  --> a.sol:11:11:
   |
11 |           Base b = u;  //~ ERROR: mismatched types
   |           ^^^^^^^^^^
```